### PR TITLE
Re-land manifest path trie update (v8)

### PR DESCRIPTION
**Pull request by jgraham**
This unbreaks the runner under the new manifest format, and makes it somewhat more stable to future changes.

### CODE REVIEW

File: tools/runner/update_manifest.py [hunk 1]
```diff
@@ -7,21 +7,31 @@
...
-        pass
-    if manifest_file is None:
-        manifest_file = manifest.Manifest("/")
-
-    update.update(root, manifest_file)
-
-    manifest.write(manifest_file, path)
-
-    return [("Content-Type", "application/json")], json.dumps({"url": "/MANIFEST.json"})
+    # TODO make this download when required
```
Comment (stephenmcgruer) [inline tools/runner/update_manifest.py]: Is this implying that:

- this always downloads, even when it shouldn't, or
- this never downloads, but should sometimes

(I'm not familiar with what 'downloads' means here)

File: tools/runner/update_manifest.py [hunk 2]
```diff
@@ -7,21 +7,31 @@
...
+    manifest_file = manifest.load_and_update(root, path, "/", parallel=False)
+
+    data = {"items": {},
+            "url_base": "/"}
+    for item_type, path, tests in manifest_file.itertypes():
+        if item_type not in ["testharness", "reftest", "manual"]:
```
Comment (stephenmcgruer) [inline tools/runner/update_manifest.py]: Could this not just pass `["testharness", "reftest", "manual"]` to `manifest_file.itertypes()`? (I may be missing something)

File: tools/runner/update_manifest.py [hunk 3]
```diff
@@ -7,21 +7,31 @@
...
+            continue
+        if item_type not in data["items"]:
```
Comment (stephenmcgruer) [inline tools/runner/update_manifest.py]: Is it expected that `itertypes()` can return the same `item_type` more than once?

**Review by stephenmcgruer — APPROVED** [review]
Some new-to-this questions, but generally LGTM as a fix.

File: tools/runner/update_manifest.py (hunk 3 above)
Comment (jgraham) [reply tools/runner/update_manifest.py]: Yes, although I changed the code to avoid this conditional.

File: tools/runner/update_manifest.py (hunk 2 above)
Comment (jgraham) [reply tools/runner/update_manifest.py]: Good point.

File: tools/runner/update_manifest.py (hunk 1 above)
Comment (jgraham) [reply tools/runner/update_manifest.py]: "This never downloads, but should, sometimes". I updated the comment a little.

**Review by stephenmcgruer — APPROVED** [review]
Still LGTM

**Comment by stephenmcgruer** [comment]
Due to the (required) use of 'parallel=False', this needs to land alongside the new manifest code. @jgraham has asked me to finish landing this.

My plan is to add re-lands of the manifest changes to this PR, then (once the tests are green), land it all together as a rebase + merge. That will preserve the series of commits and thus the history. In theory one could check out the middle of the commits and be in a broken state, but I consider that user error :)

**Comment by stephenmcgruer** [comment]
Testing locally, this appears to be working as far as I can tell.

**Pull request merged**